docs/proposals/device-crd.md

```diff
@@ -0,0 +1,670 @@
+|-------------------------------|---------------------------------------------- |
+|New Device Model Created       |  NA                                           |
+|New Device Created             | The controller creates a new config map to store the device properties and visitors defined in the device model associated with the device.  This config map is stored in etcd. The existing config map sync mechanism in the edge controller is used to sync the config map to the egde. The mapper application running in a container can get the updated config map and use the property and visitor metadata to access the device. The device controller additionally reports the device twin metadata updates to the edge node.|
+|Device Node Membership Updated | The device controller sends a membership update event to the edge node.|
+|Device  Twin Desired State Updated | The device controller sends a twin update event to the edge node.|
+|Device Model Updated           |  TODO: What happens to existing devices using this model which are generating telemetry data as per the old model ? Do we update the config map associated with the devices which are using this device model ?|
+|Device Deleted                 | The controller sends the device twin delete event to delete all device twins associated with the device. It also deletes config maps associated with the device and this delete event is synced to the edge. The mapper application effectively stops operating on the device.|
+|Device Model Deleted           |  The controller needs to run [`finalizers`](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#finalizers) to ensure that all device instances using the device model are deleted first, and only then should the model deletion proceed.|
+
+The idea behind using config map to store device properties and visitors is that these metadata are only required by the mapper applications running on the edge node in order to connect to the device and collect data.
+Mappers if run as containers can load these properties as config maps . Any additions , deletions or updates to properties , visitors etc in the cloud are watched upon by the downstream controller and config maps are updated in etcd. The existing edge controller already has the mechanism to watch on config map updates and push them to the edge node. A mapper application can get these updates and then adjust the data collection process.
+A separate design proposal can be prepared to illustrate the details of how mappers can leverage these config maps.
+
+A sample config map for the device model described earlier is shown below
+
+### Device Config Map sample
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: device1-config
+  namespace: foo
+data:
+  deviceProfile.json: |-
+    {
+      "deviceInstances": [
+        {
+          "id": "1",
+          "name": "device1",
+          "protocol": "ble-01", // TODO: not supported in API definition
```

Comment:
It's actually the name of protocolConfig, and only needed when converting from devices and deviceModels to the mappers' config. You can make it generated by controller.

Reply:
ok got it.